habitat-lab/habitat/tasks/rearrange/actions/actions.py

```diff
@@ -386,6 +386,290 @@ def step(self, *args, is_last_action, **kwargs):
             return {}
 
 
+class Controller:
+    """
+        A controller that takes the input of the waypoints and
+        produces the velocity command.
+    """
+    def __init__(self, v_max_default, w_max_default, track_yaw=True):
+        # If we want track yaw or not
+        self.track_yaw = track_yaw
+
+        # Parameter of the controller
+        self.v_max = v_max_default
+        self.w_max = w_max_default
+
+        self.lin_error_tol = self.v_max / 120
+        self.ang_error_tol = self.w_max / 120
+
+        self.acc_lin = 2.4
+        self.acc_ang = 2.4
+
+        # Initialize the parameters
+        self.base_pose_goal = np.zeros(3)
+        self.base_velocity_goal = np.zeros(3)
+
+    def set_goal(self, goal, vel_goal=None):
+        """
+        Set the goal of the agent
+        """
+        self.base_pose_goal = goal
+        if vel_goal is not None:
+            self.base_velocity_goal = vel_goal
+
+    def _compute_error_pose(self, base_pose, sim=None):
+        """
+        Updates error based on robot localization
+        """
+        base_pose_err = self._transform_global_to_base(self.base_pose_goal, base_pose, sim)
+
+        if not self.track_yaw:
+            base_pose_err[2] = 0.0
+
+        return base_pose_err
+
+    @staticmethod
+    def _velocity_feedback_control(base_pose_err, a, v_max):
+        """
+        Computes velocity based on distance from target.
+        Used for both linear and angular motion.
+
+        Current implementation: Trapezoidal velocity profile
+        """
+        t = np.sqrt(2.0 * abs(base_pose_err) / a)
+        v = min(a * t, v_max)
+        return v * np.sign(base_pose_err)
+
+    @staticmethod
+    def _turn_rate_limit(lin_err, heading_diff, w_max, tol=0.0):
+        """
+        Compute velocity limit that prevents path from overshooting goal
+
+        heading error decrease rate > linear error decrease rate
+        (w - v * np.sin(phi) / D) / phi > v * np.cos(phi) / D
+        v < (w / phi) / (np.sin(phi) / D / phi + np.cos(phi) / D)
+        v < w * D / (np.sin(phi) + phi * np.cos(phi))
+
+        (D = linear error, phi = angular error)
+        """
+        assert lin_err >= 0.0
+        assert heading_diff >= 0.0
+
+        max_heading_ang = np.pi / 10
```

Comment:
put hard-coded values like this at the top of the file please, we can set them in consturctor later

Reply:
Since this is a static method, I think this should be checked inside the function.